syntax = "proto3";

package esb.agent.v1;
option go_package = "github.com/poruru/edge-serverless-box/services/agent/pkg/api/v1";

service AgentService {
  // コンテナを確保し、接続情報を返す (なければ起動、あれば再利用)
  rpc EnsureContainer (EnsureContainerRequest) returns (WorkerInfo);
  
  // 明示的にコンテナを停止・削除する
  rpc DestroyContainer (DestroyContainerRequest) returns (DestroyContainerResponse);

  // コンテナを一時停止する (Warm Start 用)
  rpc PauseContainer (PauseContainerRequest) returns (PauseContainerResponse);

  // コンテナを再開する (Warm Start 用)
  rpc ResumeContainer (ResumeContainerRequest) returns (ResumeContainerResponse);

  // 管理下の全コンテナの状態を取得 (Phase 3: Janitor 用)
  rpc ListContainers (ListContainersRequest) returns (ListContainersResponse);
}

message PauseContainerRequest {
  string container_id = 1;
}

message PauseContainerResponse {
  bool success = 1;
}

message ResumeContainerRequest {
  string container_id = 1;
}

message ResumeContainerResponse {
  bool success = 1;
}

message EnsureContainerRequest {
  string function_name = 1;
  string image = 2;
  map<string, string> env = 3;
}

message DestroyContainerRequest {
  string function_name = 1;
  string container_id = 2;
}

message DestroyContainerResponse {
  bool success = 1;
}

message WorkerInfo {
  string id = 1;
  string name = 2;
  string ip_address = 3;
  int32 port = 4;
}

// Phase 3: ListContainers 用メッセージ
message ListContainersRequest {}

message ListContainersResponse {
  repeated ContainerState containers = 1;
}

message ContainerState {
  string container_id = 1;
  string function_name = 2;
  string status = 3;      // "RUNNING", "PAUSED", "STOPPED", "UNKNOWN"
  int64 last_used_at = 4; // Unix Timestamp (秒)
}
